test/e2e/pvutil.go

```diff
@@ -0,0 +1,597 @@
+
+	// Re-get the pv and pvc objects
+	pv, err = c.Core().PersistentVolumes().Get(pv.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	// Re-get the pvc and
+	pvc, err = c.Core().PersistentVolumeClaims(ns).Get(pvc.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	// The pv and pvc are both bound, but to each other?
+	// Check that the PersistentVolume.ClaimRef matches the PVC
+	Expect(pv.Spec.ClaimRef).NotTo(BeNil())
+	Expect(pv.Spec.ClaimRef.Name).To(Equal(pvc.Name))
+	Expect(pvc.Spec.VolumeName).To(Equal(pv.Name))
+	Expect(pv.Spec.ClaimRef.UID).To(Equal(pvc.UID))
+}
+
+// Search for bound PVs and PVCs by examining pvols for non-nil claimRefs.
+// NOTE: Each iteration waits for a maximum of 3 minutes per PV and, if the PV is bound,
+//   up to 3 minutes for the PVC. When the number of PVs != number of PVCs, this can lead
+//   to situations where the maximum wait times are reached several times in succession,
+//   extending test time. Thus, it is recommended to keep the delta between PVs and PVCs
+//   small.
+func waitAndVerifyBinds(c clientset.Interface, ns string, pvols pvmap, claims pvcmap, testExpected bool) {
+
+	var actualBinds int
+	expectedBinds := len(pvols)
+	if expectedBinds > len(claims) { // want the min of # pvs or #pvcs
+		expectedBinds = len(claims)
+	}
+
+	for pvName := range pvols {
+		err := framework.WaitForPersistentVolumePhase(v1.VolumeBound, c, pvName, 3*time.Second, 180*time.Second)
+		if err != nil && len(pvols) > len(claims) {
+			framework.Logf("WARN: pv %v is not bound after max wait", pvName)
+			framework.Logf("      This may be ok since there are more pvs than pvcs")
+			continue
+		}
+		Expect(err).NotTo(HaveOccurred())
+
+		pv, err := c.Core().PersistentVolumes().Get(pvName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		if cr := pv.Spec.ClaimRef; cr != nil && len(cr.Name) > 0 {
+			// Assert bound pvc is a test resource. Failing assertion could
+			// indicate non-test PVC interference or a bug in the test
+			pvcKey := makePvcKey(ns, cr.Name)
+			_, found := claims[pvcKey]
+			Expect(found).To(BeTrue())
+
+			err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, c, ns, cr.Name, 3*time.Second, 180*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+			actualBinds++
+		}
+	}
+
+	if testExpected {
+		Expect(actualBinds).To(Equal(expectedBinds))
+	}
+}
+
+// Test the pod's exit code to be zero.
+func testPodSuccessOrFail(c clientset.Interface, ns string, pod *v1.Pod) {
+
+	By("Pod should terminate with exitcode 0 (success)")
+	err := framework.WaitForPodSuccessInNamespace(c, pod.Name, ns)
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("Pod %v succeeded ", pod.Name)
+}
+
+// Deletes the passed-in pod and waits for the pod to be terminated. Resilient to the pod
+// not existing.
+func deletePodWithWait(f *framework.Framework, c clientset.Interface, pod *v1.Pod) {
+
+	if pod == nil {
+		return
+	}
+	framework.Logf("Deleting pod %v", pod.Name)
```

Comment:
@liggitt @saad-ali is there a label that can be applied to re-run the build/test? I looked at the 3 failures below and, unless I'm missing something, they appear to be build/test infra related. Thanks.